### PR TITLE
Edge security: Communication between IOT Devices and Edges

### PR DESCRIPTION
**Pull request by pradumnnec**

### CODE REVIEW

**Review by showersky — APPROVED** [review]

File: docker/edge/http/script.sh [hunk 1]
```diff
@@ -45,41 +43,46 @@ do
...
+
+ID=`cat API_detail.js | tail -1| jq -r '.[] | .[] | .id' | tr -d '\n'`
+
+
+#To fetch secret of the Application 
+for connect2 in 1 2 3 
+do 
+  curl --include \
+     --header "X-Auth-token: $token" \
+  "http://$IP_ADDRESS:3000/v1/applications/$ID" > accessAPI.txt
```
Comment (naveennec) [inline docker/edge/http/script.sh:76]: [Suggestion] accessAPI.txt is a hidden file better to make it hidden file.

File: docker/edge/http/stop.sh [hunk 2]
```diff
@@ -1,10 +1,7 @@
-sh $(pwd)/delete.sh
-if [ $? -eq 0 ]; then
-    docker stop metricbeat && docker rm $_
-    docker stop edgebroker && docker rm $_
-    docker stop edgeworker && docker rm $_
-else
-     echo failed in delete Application
-fi
+docker stop metricbeat && docker rm $_
```
Comment (naveennec) [inline docker/edge/http/stop.sh:1]: [Minor] Write proper message in front of this file -- why this file is created for.

File: docker/edge/http/start.sh [hunk 3]
```diff
@@ -7,8 +7,11 @@ fi
 sh $(pwd)/script.sh
 if [ $? -eq 0 ]; then
     docker run -d   --name=metricbeat   --user=root   --volume="$(pwd)/metricbeat.docker.yml:/usr/share/metricbeat/metricbeat.yml:ro"   --volume="/var/run/docker.sock:/var/run/docker.sock:ro"   --volume="/sys/fs/cgroup:/hostfs/sys/fs/cgroup:ro"   --volume="/proc:/hostfs/proc:ro"   --volume="/:/hostfs:ro"   docker.elastic.co/beats/metricbeat:7.6.0 metricbeat -e   -E output.elasticsearch.hosts=["<Cloud_Public_IP>:9200"]
-    docker run -d --name=edgebroker -v $(pwd)/config.json:/config.json -p 8170:8170  fogflow/broker:$htype
+    docker run -d --name=edgebroker -v $(pwd)/config.json:/config.json -p 8060:8060  fogflow/broker:3.2
```
Comment (naveennec) [inline docker/edge/http/start.sh:10]: [minor]type should be configurable not constant.

File: docker/edge/http/start.sh [hunk 4]
```diff
@@ -7,8 +7,11 @@ fi
...
     docker run -d --name=edgeworker -v $(pwd)/config.json:/config.json -v /tmp:/tmp -v /var/run/docker.sock:/var/run/docker.sock fogflow/worker:$htype
+    docker run -d --name=pepEdge -v $(pwd)/pep-config.js:/opt/fiware-pep-proxy/config.js -p 5556:5556 fiware/pep-proxy
+    #sh $(pwd)/user_addition.sh
+    #docker run -d --name=Nginx -v $(pwd)/nginx.conf:/etc/nginx/nginx.conf -v $(pwd)/user:/etc/nginx/user -p 80:80 nginx
```
Comment (naveennec) [inline docker/edge/http/start.sh:14]: [minor] remove commented code

File: docker/edge/http/start.sh [hunk 5]
```diff
@@ -7,8 +7,11 @@ fi
...
 else
       echo failed security check
```
Comment (naveennec) [inline docker/edge/http/start.sh:16]: [Suggestion] Better to enclose it in inverted commas.

File: docker/edge/http/script.sh [hunk 6]
```diff
@@ -98,13 +152,17 @@ if [ $? -eq 0 ]; then
...
-echo "end of script"
-rm -rf accessAPI.txt access_token.txt generate_token.txt
-
+#To generate Authorization Token
 
+AUTH=`echo -n $ID:$SECRET | base64 | tr -d "\t\r\n"` 
+echo ------------------------------------------
+echo "Authorization : Basic "$AUTH 
+echo ------------------------------------------
+echo "end of script"
```
Comment (naveennec) [inline docker/edge/http/script.sh:167]: [minor] Better to explain about which script ended. like authentication script.

File: docker/edge/http/script.sh [hunk 7]
```diff
@@ -88,7 +91,58 @@ echo -----------------------------
...
+
+#setting up of pep-config.js
+sed -i "s/PEP_PROXY_IDM_HOST.*/PEP_PROXY_IDM_HOST \|\| '${IP_ADDRESS}',/" pep-config.js
+sed -i "s/PEP_PROXY_APP_HOST.*/PEP_PROXY_APP_HOST \|\| '${EDGE_IP_ADDRESS}',/" pep-config.js
+sed -i "s/PEP_PROXY_APP_PORT.*/PEP_PROXY_APP_PORT \|\| '8060',/" pep-config.js
+sed -i "s/PEP_PROXY_APP_ID.*/PEP_PROXY_APP_ID \|\| '${ID}',/" pep-config.js
+sed -i "s/PEP_PROXY_USERNAME.*/PEP_PROXY_USERNAME \|\| '${Pep_ID}',/" pep-config.js 
+sed -i "s/PEP_PASSWORD.*/PEP_PASSWORD \|\| '${Pep_password}',/" pep-config.js
+
+
```
Comment (naveennec) [inline docker/edge/http/script.sh:144]: [minor] Explain why this for loop is used.

File: docker/edge/http/script.sh [hunk 8]
```diff
@@ -88,7 +91,58 @@ echo -----------------------------
...
+     --header "X-Auth-token: $token" \
+  "http://$IP_ADDRESS:3000/v1/applications/$ID/pep_proxies" > PEP_Detail.js
+   if [ $? -eq 0 ]; then
+      break
+   else
+      echo reconnecting...
+      continue
+    fi
+done
+
```
Comment (naveennec) [inline docker/edge/http/script.sh:107]: [Major] what about the scenario when PEP Proxy didn't connected still PEP_Details.js will be read. Need to write the logic.

File: docker/edge/http/script.sh [hunk 9]
```diff
@@ -88,7 +91,58 @@ echo -----------------------------
...
+     --header "X-Auth-token: $token" \
+  "http://$IP_ADDRESS:3000/v1/applications/$ID/pep_proxies" > PEP_Details.txt 
+  if [ $? -eq 0 ]; then
+     break
+  else
+     echo reconnecting...
+     continue
+  fi
+done
+
```
Comment (naveennec) [inline docker/edge/http/script.sh:126]: [Major] what about the scenario when PEP Proxy didn't gets connected and so password not generated so better to write the hidden handling.

File: docker/edge/http/pep-config.js [hunk 10]
```diff
@@ -0,0 +1,81 @@
...
+config.https = {
+  enabled: toBoolean(process.env.PEP_PROXY_HTTPS_ENABLED, false),
+  cert_file: 'cert/cert.crt',
+  key_file: 'cert/key.key',
+  port: process.env.PEP_PROXY_HTTPS_PORT || 443,
+};
+
+//IDM configration
+config.idm = {
+  host: process.env.PEP_PROXY_IDM_HOST || '180.179.214.199',
```
Comment (naveennec) [inline docker/edge/http/pep-config.js:24]: [Suggestion] why these i.p are hard coded why not configured in a configuration file then read here.

File: docker/edge/http/pep-config.js [hunk 11]
```diff
@@ -0,0 +1,81 @@
...
+};
+
+config.organizations = {
+  enabled: toBoolean(process.env.PEP_PROXY_ORG_ENABLED, false),
+  header: process.env.PEP_PROXY_ORG_HEADER || 'fiware-service'
+}
+
+// Credentials obtained when registering PEP Proxy in app_id in Account Portal
+config.pep = {
+  app_id: process.env.PEP_PROXY_APP_ID || '6e396def-3fa9-4ff9-84eb-266c13e93964',
```
Comment (naveennec) [inline docker/edge/http/pep-config.js:43]: [Suggestion] Hardcoded code must be replaced with configurable one.

File: docker/edge/http/script.sh [hunk 12]
```diff
@@ -45,41 +43,46 @@ do
...
 
+#cat generate_token.txt
+
 token=`grep "X-Subject-Token" generate_token.txt | cut -f 2 -d ":" |  sed 's/\r$//g' | tr -d '\n'`
 echo -----------------------------
 echo IDM token is $token 
 echo -----------------------------
 
-#Register application, it will return Application ID and Password. ID and Password will use to genenrate access token
+#Fetch Application ID and Password. ID and Password will use to genenrate access token
```
Comment (naveennec) [inline docker/edge/http/script.sh:53]: [Minor]Fetch application ID and shared secret to authenticate the token shared by IOT devices.

File: docker/edge/http/script.sh [hunk 13]
```diff
@@ -45,41 +43,46 @@ do
...
-          \"jwt\",
-          \"permanent\"
-      ]
-    }
-  }" \
-  "http://$IP_ADDRESS:3000/v1/applications" > accessAPI.txt
+  curl --include \
+     --request GET \
+     --header "X-Auth-token: $token" \
+  "http://$IP_ADDRESS:3000/v1/applications" > API_detail.js
```
[the remark this answers is not in the document]
Comment (pradumnnec) [reply docker/edge/http/script.sh:59]: Comment Incorporated

File: docker/edge/http/script.sh (hunk 1 above)
Comment (pradumnnec) [reply docker/edge/http/script.sh:76]: Comment Incorporated

File: docker/edge/http/stop.sh (hunk 2 above)
Comment (pradumnnec) [reply docker/edge/http/stop.sh:1]: Comment Incorporated

File: docker/edge/http/start.sh (hunk 3 above)
Comment (pradumnnec) [reply docker/edge/http/start.sh:10]: Type is configurable in the script

File: docker/edge/http/start.sh (hunk 4 above)
Comment (pradumnnec) [reply docker/edge/http/start.sh:14]: Commented code is removed

File: docker/edge/http/start.sh (hunk 5 above)
Comment (pradumnnec) [reply docker/edge/http/start.sh:16]: commented code is removed

File: docker/edge/http/script.sh (hunk 6 above)
Comment (pradumnnec) [reply docker/edge/http/script.sh:167]: Incorporated proper comment

File: docker/edge/http/script.sh (hunk 7 above)
Comment (pradumnnec) [reply docker/edge/http/script.sh:144]: this is used for configuring pep-proxy

File: docker/edge/http/script.sh (hunk 8 above)
Comment (pradumnnec) [reply docker/edge/http/script.sh:107]: Added code for handling such event

File: docker/edge/http/pep-config.js (hunk 10 above)
Comment (pradumnnec) [reply docker/edge/http/pep-config.js:24]: these ip are automatically configured once the edge is started.

File: docker/edge/http/pep-config.js (hunk 11 above)
Comment (pradumnnec) [reply docker/edge/http/pep-config.js:43]: It is not hard coded, rather updated while  bringing the up the edge

File: docker/edge/http/script.sh (hunk 12 above)
Comment (pradumnnec) [reply docker/edge/http/script.sh:53]: Added the said comment in script

**Pull request merged**